lib/ai/model/agent.go

```diff
@@ -0,0 +1,319 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package model
+
+import (
+	"context"
+	"encoding/json"
+	"strings"
+	"time"
+
+	"github.com/gravitational/trace"
+	"github.com/sashabaranov/go-openai"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	actionFinalAnswer = "Final Answer"
+	actionException   = "_Exception"
+	maxIterations     = 15
+	maxElapsedTime    = 5 * time.Minute
+)
+
+var AssistAgent = &Agent{
+	tools: []Tool{
+		&commandExecutionTool{},
+	},
+}
+
+type Agent struct {
+	tools []Tool
+}
+
+type AgentAction struct {
+	action string
+	input  string
+	log    string
+}
+
+type AgentFinish struct {
+	// output must be Message or CompletionCommand
+	output any
+}
+
+type executionState struct {
+	llm               *openai.Client
+	chatHistory       []openai.ChatCompletionMessage
+	humanMessage      openai.ChatCompletionMessage
+	intermediateSteps []AgentAction
+	observations      []string
+	tokensUsed        *TokensUsed
+}
+
+func (a *Agent) Think(ctx context.Context, llm *openai.Client, chatHistory []openai.ChatCompletionMessage, humanMessage openai.ChatCompletionMessage) (any, error) {
+	log.Debug("entering agent think loop")
+	iterations := 0
+	start := time.Now()
+	shouldExit := func() bool { return iterations > maxIterations || time.Since(start) > maxElapsedTime }
+	tokensUsed := newTokensUsed_Cl100kBase()
+	state := &executionState{
+		llm:               llm,
+		chatHistory:       chatHistory,
+		humanMessage:      humanMessage,
+		intermediateSteps: make([]AgentAction, 0),
+		observations:      make([]string, 0),
+		tokensUsed:        tokensUsed,
+	}
+
+	for {
+		log.Debugf("performing iteration %v of loop, %v seconds elapsed", iterations, int(time.Since(start).Seconds()))
+
+		// This is intentionally not context-based, as we want to finish the current step before exiting
+		// and the concern is not that we're stuck but that we're taking too long over multiple iterations.
+		if shouldExit() {
+			return nil, trace.Errorf("timeout: agent took too long to finish")
```

Comment:
I think that one of the names is misleading. The function suggests that the loop should be exited, but the error message says that we hit a timeout. Can we rename the function or adjust the returned error (probably the first one as the more detailed error message is more useful).

Reply:
Done